clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionEventSimulationTest.java

```diff
@@ -0,0 +1,147 @@
+
+    @Before
+    public void setup() {
+        transactionManager = new TransactionManager(logContext, "txn-id",
+            requestTimeoutMs, apiVersion, new ApiVersions(), false);
+        transactionCoordinator = new TransactionSimulationCoordinator(client);
+    }
+
+    @Test
+    public void simulateTxnEvents() throws InterruptedException {
+        final int batchSize = 100;
+        final int lingerMs = 0;
+        final int deliveryTimeoutMs = 10;
+
+        RecordAccumulator accumulator = new RecordAccumulator(logContext, batchSize, CompressionType.GZIP,
+            lingerMs, retryBackOffMs, deliveryTimeoutMs, new Metrics(), "accumulator", time, new ApiVersions(), transactionManager,
+            new BufferPool(1000, 100, new Metrics(), time, "producer-internal-metrics"));
+
+        metadata.add("topic", time.milliseconds());
+        metadata.update(metadata.newMetadataRequestAndVersion(time.milliseconds()).requestVersion,
+            TestUtils.metadataUpdateWith(1, Collections.singletonMap("topic", 2)), true, time.milliseconds());
+
+        sender = new Sender(logContext, client, metadata, accumulator, false, 100, (short) 1,
+            Integer.MAX_VALUE, new SenderMetricsRegistry(new Metrics()), time, requestTimeoutMs, 10, transactionManager, new ApiVersions());
+
+        transactionManager.initializeTransactions();
+        sender.runOnce();
+        resolvePendingRequests();
+        final int numTransactions = 100;
+
+        TopicPartition key = new TopicPartition("topic", 0);
+        long committedOffsets = 0L;
+        Random abortTxn = new Random();
+        client.prepareMetadataUpdate(TestUtils.metadataUpdateWith(1, Collections.singletonMap("topic", 2)));
+        final long timestamp = 0L;
+        final int maxBlockTime = 0;
+
+        for (int i = 0; i < numTransactions; i++) {
+            transactionManager.beginTransaction();
+            transactionManager.maybeAddPartitionToTransaction(key);
+            accumulator.append(key, timestamp, new byte[1], new byte[1],
+                Record.EMPTY_HEADERS, null, maxBlockTime, false, time.milliseconds());
+            transactionManager.sendOffsetsToTransaction(
+                Collections.singletonMap(key, new OffsetAndMetadata(committedOffsets)),
+                new ConsumerGroupMetadata("group"));
+
+            if (abortTxn.nextBoolean()) {
+                transactionManager.beginCommit();
+                committedOffsets += 1;
+            } else {
+                transactionManager.beginAbort();
+            }
+
+            resolvePendingRequests();
+        }
+
+        assertTrue(transactionCoordinator.persistentPartitionData().containsKey(key));
+        assertTrue(transactionCoordinator.committedOffsets().containsKey(key));
+        assertEquals(committedOffsets - 1, (long) transactionCoordinator.committedOffsets().get(key));
+    }
+
+    private void resolvePendingRequests() {
+        Random dropMessageRandom = new Random();
```

Comment:
It's useful if the simulation test is deterministic. That way failures are easy to reproduce. Perhaps we can use a shared `Random` instance (between this class and the coordinator) with a defined seed.

Reply:
If we have a defined seed, nextRandom will be always pointing at either true or false. We may consider using a `nextInt() %2` thing